src/shared/logger.ts

```diff
@@ -17,16 +17,18 @@ import { DefaultSettingsConfiguration, SettingsConfiguration } from './settingsC
 const localize = nls.loadMessageBundle()
 
 const LOG_RELATIVE_PATH: string = path.join('Code', 'logs', 'aws_toolkit')
-const DEFAULT_LOG_LEVEL: string = 'info'
+const DEFAULT_LOG_LEVEL: LogLevel = 'info'
 const DEFAULT_LOG_NAME: string = `aws_toolkit_${makeDateString('filename')}.log`
 const DEFAULT_OUTPUT_CHANNEL: vscode.OutputChannel = vscode.window.createOutputChannel('AWS Toolkit Logs')
 
 let defaultLogger: Logger
 
+export type LogLevel = 'debug' | 'verbose' | 'info' | 'warn' | 'error'
```

Comment:
Do we want to provide our own enums or use the Winston-provided ones? You should be able to check this with something like `Object.keys(logger.levels)`. See the writeToLogs function (currently at  https://github.com/aws/aws-toolkit-vscode/blob/develop/src/shared/logger.ts#L175 ...this includes a key:value pairing to a number that can be used for level comparisons)

Reply:
I do think we should use a type more specific than string, hence `LogLevel`. I'm not sure whether it would be a good idea to surface any Winston specific implementation out side of `shared/logger`. Ideally consumers of `shared/logger` need not know anything about Winston and we would be able to swap to another logger without affecting any code outside of `shared/logger`.

In all of the cases I've used `logLevel`, it has been the string that I needed, thus a string enum, as I mentioned earlier in our logger discussions would be useful for both type safety and displaying string values. The number associated with a traditional enum would not be useful for anything I've done so far outside of `shared/logger`.

Regarding LogLevel, any idea why we might use `verbose`? I'm used to using console which does not provide this option but does provide `trace`. I asked another team whether they used `verbose`, and the looked puzzled then asked "Do you mean `trace`?" We may wish to consider using the more standard `trace` or simply excluding `verbose`.

Given the choices:

1. Tying ourselves exactly to Winston's full choices of log levels
2. Allowing any old string to be used as a log level
3. Taking a few seconds of time to define our own type

I'd prefer option 3.

Reply:
BTW, I do like/prefer the idea of log levels stemming from a single source and like the idea of your suggestion `Object.keys(logger.levels)`. Unfortunately, do to the way our logger is initialized, `defaultLogger`, is not assigned at `shared/logger` initialization time but instead some time in the future when `extension.ts` > `activate` is called. Thus the types would be defined at runtime and rathe than compile time and are not available for compiler verification purposes.

Here are two solutions I see for LogLevel to deprive from a single source of truth:

1. Adjust `shared/logger` to assign `defaultLogger` during module initialization and use `Object.keys(logger.levels)` if that works.
2. Define an object or string enum that defines the log levels at package initialization time and is used to:

- configure the Winston logger
- provide compile time verification of logLevel type within `shared/logger`
- export the LogLevel type for external type verification

This need not be done at this time. Perhaps if create an associated issue.

Either way it would seem odd to ague that the current change to which you refer:

- using a `LogLevel` type to provide some compile time verification for `logLevel`

Is a small step forward from providing no compile time type verification for `logLevel`.

Reply:
We can define our own type that is usable by Winston: https://github.com/winstonjs/winston#using-custom-logging-levels , if you'd rather do this, we can assign it on log creation (for both the default and any custom loggers we build). If you do this, be sure to change our configured enums in the package.json file (currently at https://github.com/aws/aws-toolkit-vscode/blob/develop/package.json#L68 ).

Since we have levels configured in package.json, we won't ever be able to pull from a single source of truth (unless we can access the package.json file in our code?).

One note, the enum will only enforce on calling `logger.log('level', stuff, to, log, ... )`. I prefer not using this and using explicit calls to `.info`, `.warn`, `.error`, etc. In that case, our exposed functions are our _de facto_ enums.

Also, changing to initialize at module load instead of during plugin activation will cause issues with our current testing implementation; right now, we load a logger prior to each test so we can clean it up afterwards (otherwise, I don't think we have a concept of `runAfterEveryTestHasCompleted`. For what it's worth, I'm a fan of on module load (since that way we don't have to call `getLogger` in every single function and class that logs)...I'd definitely be interested if you can find a way to do this that plays well with tests.

Reply:
Agreed, if we refactor this I'd prefer option 2, define our own enum/type/object to be used by Winston. 

We could address module initialization at a later time. I do think there is a good way to auto-initialize `shared/logger` that plays well with testing but that's a separate issue.

Reply:
✔️